### PR TITLE
gnrc_tcp: return immediatly on gnrc_tcp_recv if a connection is closing

### PR DESCRIPTION
**Pull request by brummer-simon**
This PR fixes the stale Issue mentioned in this PR #10899. I think the original Author abandoned his PR, but the Issue remains till today. To summarize the Issue the behaviour of gnrc_tcp_recv() was not fully in line with the [TCP RFC](https://tools.ietf.org/html/rfc793).

Specified Behaviour: 
gnrc_tcp_recv() must return immediately if a packet with the FIN Bit set, was received from the peer indicating that want to close a connection. As a response a call to gnrc_tcp_recv() return all queued data up to this point immediately, a given timeout must be ignored because there will be no new data to receive.

Actual Behaviour without this PR:
gnrc_recv() will block even if a packet with the FIN Bit set, was received, if a timeout was given.
Additionally the user of gnrc_tcp has no means to figure out that the peer wants to close the connection.

A more detailed discussion on the Issue can be found in the original PR  #10899.

### CODE REVIEW

**Comment by miri64** [comment]
Can you give both your commit and the PR a bit more human readable summary? ;-)

**Comment by brummer-simon** [comment]
> Can you give both your commit and the PR a bit more human readable summary? ;-)

Better now?

**Comment by brummer-simon** [comment]
@miri64: If this #12369 gets merged before this PR ping me because that will be test script no 6 then.

File: sys/include/net/gnrc/tcp.h [hunk 1]
```diff
@@ -160,6 +160,7 @@ ssize_t gnrc_tcp_send(gnrc_tcp_tcb_t *tcb, const void *data, const size_t len,
  *                                           @p user_timeout_duration_us microseconds passed.
  *
  * @returns   The number of bytes read into @p data.
+ *            Zero if the connection is closing and no further data can be read.
```
Comment (miri64) [inline sys/include/net/gnrc/tcp.h]: Document numeric value, please

```suggestion
 *            0, if the connection is closing and no further data can be read.
```

**Comment by brummer-simon** [comment]
I have changed the test index number to 6 because I think #12369 will be merged before this.

File: sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c [hunk 2]
```diff
@@ -307,9 +307,11 @@ static int _fsm_call_recv(gnrc_tcp_tcb_t *tcb, void *buf, size_t len)
...
     size_t rcvd = ringbuffer_get(&(tcb->rcv_buf), buf, len);
 
-    /* If receive buffer can store more than GNRC_TCP_MSS: open window to available buffer size */
-    if (ringbuffer_get_free(&tcb->rcv_buf) >= GNRC_TCP_MSS) {
-        tcb->rcv_wnd = ringbuffer_get_free(&(tcb->rcv_buf));
+    /* Reopen window if recv buffer can hold a MSS sized payload and FIN was not received. */
+    uint16_t buf_free = ringbuffer_get_free(&tcb->rcv_buf);
+
+    if (buf_free >= GNRC_TCP_MSS && tcb->state != FSM_STATE_CLOSE_WAIT) {
+        tcb->rcv_wnd = buf_free;
```
Comment (miri64) [inline sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c]: This change seems to be unrelated (i.e. it should at least go into a separate commit) and I'm not sure if it is valid. Is it guaranteed that `tcb->rcv_buf`does not change between the `if` and the assignment to `tcb->rcv_wnd`? Or asked differently: is the calling thread to `ringbuffer_get_free()` the same that writes to the ring buffer?

File: tests/gnrc_tcp/README.md [hunk 3]
```diff
@@ -18,6 +18,11 @@ in the tests directory.
     This test covers receiving of a byte stream from the host system. The received data is
     causing window opening and closing as well as data transmission over multiple packets.
 
+6) 06-receive_data_closed_conn.py
+    This test covers accessing received data after receiving a FIN packet. If the connection was closed
+    by the peer, a call to gnrc_tcp_recv must return directly with all currently received data
+    or zero if there is no data. The function must return immediatly dispite any given timeout.
```
Comment (miri64) [inline tests/gnrc_tcp/README.md]: Please put the new tests into a separate commit.

File: sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c (hunk 2 above)
Comment (brummer-simon) [reply sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c]: The calling threads are different, but all calls to the FSM are synchronized via a mutex. 
There is no problem there.

File: tests/gnrc_tcp/README.md (hunk 3 above)
Comment (brummer-simon) [reply tests/gnrc_tcp/README.md]: Why. Does it not make sense to supply a test that verifies a change within the same PR?

File: sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c (hunk 2 above)
Comment (brummer-simon) [reply sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c]: Separate PR: #12379

File: tests/gnrc_tcp/README.md (hunk 3 above)
Comment (miri64) [reply tests/gnrc_tcp/README.md]: Same PR yes, but they should be a separate commit, as they are not part of the fix themselves.

**Comment by miri64** [comment]
Needs rebase. You may squash while doing so.

**Comment by brummer-simon** [comment]
Rebased to current master. To stabilize the tests I had to increase the timeout of the second script.

**Comment by brummer-simon** [comment]
@miri64 - Ping due to force push.

**Comment by brummer-simon** [comment]
Done

**Comment by brummer-simon** [comment]
> Please fix the commit message to something more human-readable ;-)

Done

**Comment by brummer-simon** [comment]
@miri64 - Ping

**Comment by fjmolinas** [comment]
@brummer-simon did you run the tests on non-native boards as well? I'm having issues making them pass on other boards than `native`, should that be expected? Might I be missing some configuration?

**Comment by brummer-simon** [comment]
> 
> 
> @brummer-simon did you run the tests on non-native boards as well? I'm having issues making them pass on other boards than `native`, should that be expected? Might I be missing some configuration?

Hi @fjmolinas,

The tests run on boards as well via ethos. However some boards are fast than others there might be cases where the Timeout specified in the test scripts is to short. Would you do me a favor and increase the timeout for the failing tests? If this works, would you create a PR?

Cheers Simon

**Comment by fjmolinas** [comment]
> The tests run on boards as well via ethos. However some boards are fast than others there might be cases where the Timeout specified in the test scripts is to short. Would you do me a favor and increase the timeout for the failing tests? If this works, would you create a PR?

Thanks, I'm not sure what happened, I had a lot of issue making it work, but eventually it did with the default timeout. Thanks!
